File: algorithms/java/等差数列划分/等差数列划分.java
// Source : https://leetcode.cn/problems/arithmetic-slices/
// Author : Yanan Wang
// Date   : 2023-06-24

class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        int result = 0;
        for (int i = 2; i < n; i++) {
            if (nums[i] - nums[i - 1] == nums[i - 1] - nums[i - 2]) {
                dp[i] = dp[i - 1] + 1;
            } 
            result += dp[i];
        }
        return result;
    }
}

File: algorithms/python/等差数列划分/等差数列划分.py
# Source : https://leetcode.cn/problems/arithmetic-slices/description/
# Author : Yanan Wang
# Date   : 2024-01-16

class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * n 
        ans = 0
        
        for i in range(2, n):
            if nums[i] + nums[i - 2] == nums[i - 1] * 2:
                dp[i] = dp[i - 1] + 1
            ans += dp[i]
        
        return ans
